### ISSUE TITLE
DynamicRoutingEngine on Android

### ISSUE THREAD

**Issue by StefanTiess** [heremaps/here-sdk-examples#231]
Hi, 

I want to ask you for some clarification on how to best use the DynamicRoutingEngine as the example apps are not clear about it and I see lots of native crashes and ANRs related to it.

Lifecycle of the Engine
When I start navigation, I also start dynamic routing updates with dynamicRoutingEngine.start(route, dynamicroutinglistener). I update the location with each NavigableLocationUpdate
=> How am I supposed to react if the route changes during navigation
* If there is a route deviation event fired (so the user is no longer on the precalculated route)
* if a new route was calculated with the returntoroute engine
* if a new dynamic route was calculated
* if the user adds or removes a waypoint and an entirely new route has to be calculated
=> In what cases do I need to restart the dynamic routing engine to provide the new route information
=> Do I need to stop it before I can start it again? Is there any other way to update the route that needs to be updated

Crashes
=> I regularly experience ANRs when stopping the dynamic routing engine on the main thread. Is it expected to not run the dynamicRoutingEngine from the main thread?
=> Also I regularly see native crashes in the "disposeNativeHandle" function of the dynamicroutingengine. What can cause this behavior and how can I avoid it?
=> Also the dynamicRoutingEngine logs cryptic messages before crashes, that do not help me at all. For Example:
`[ERROR] traffic-aware-navigation - updated route: Routing error 16` happened right before a crash

I would really appreciate some best practices here and better logs. Currently this engine is responsible for the majority of all crashes and ANRs, and there is little to none information on how to resolve them.

**Comment by HERE-SDK-Support-Team** [comment]
Hi @StefanTiess,

please take a look at this [guide](https://developer.here.com/documentation/android-sdk-navigate/dev_guide/topics/navigation.html#dynamically-find-better-routes) that explains how to use the dynamic routing engine. 

Then please also take a look at the example app that shows how to get notified `[here](https://github.com/heremaps/here-sdk-examples/blob/d1baba6878ea3dd89c778dadaa7be74938d4564e/examples/latest/navigate/android/Navigation/app/src/main/java/com/here/navigation/NavigationExample.java#L696)`.

DynamicRoutingEngine and deviation event are two separate things. ReturnToRoute has also nothing to do with the DynamicRoutingEngine.

As you can see in the link, `dynamicRoutingEngine.start(route, new DynamicRoutingListener(){...})` requires a route and a listener, this should clarify your question. Whenever you set a new route to the VisualNavigator, you would also like to call start() on the dynamicRoutingEngine.

Of course, this is expected to run on the main thread.

**Comment by StefanTiess** [comment]
Hi, 
thanks for your update. I am also further investigating and rebuilding the integration according to your suggestions.
Previously whenever there was a new route calculated, i immediately stopped and started the DynamicRoutingEngine with the new route. This was leading to a lot of ANRs. Now I stop as soon as a Route Deviation is detected or a new route is requested by the driver during navigation. Only after the new route is calculated and passed to the visualNavigatior I start the DynamicRouting Engine again. The only exception is a Route provided directly from the DynamicRoutingEngine. This will just be passed to the visualNavigatior without restarting the DynamicRoutingEngine.

So far, so good.

But. There is still the issue with Crashes. It seems to me the DynamicRoutingEngine is not processing RoutingErrors correctly. I see the Exception in the Logs, but the onRouting Event is either not called or the RoutingError.name is not processed as it should. In any case: Sometimes it correctly passes on the routingError, sometimes it just crashes out of the blue.

This is how it looks in the crash Logs:

`[ERROR] routing - Routing error: 3, exception: Could not resolve host: router.hereapi.com
  [ERROR] traffic-aware-navigation - updated route: Routing error 3
  A  java_vm_ext.cc:578] JNI DETECTED ERROR IN APPLICATION: non-nullable const char* was NULL
                                                                                                    java_vm_ext.cc:578]     in call to GetStaticFieldID
                                                                                                    java_vm_ext.cc:578]     from void com.here.sdk.core.threading.RunnableImpl.run()` 

The only thing i was doing in the `onRoutingError` Callback was to log the routingError. But even after removing the logging, Crashes are still appearing. 

A good way of reproducing these crashes is to start navigation and and the DynamicRoutingEngine and after some time deactivate connectivity of the test device. This is something that frequently happens to drivers, when they drive through an area of low mobile reception. At some point the dynamicRoutingEngine will report Exception 3 as the host is not reachable. But instead of properly reporting the issue as a RoutingError, it crashes without any means to intercept. 

This would need to be fixed asap, as it usually crashes the app in the middle of driving, tempting the driver to resolve the issue while driving, hence a serious safety risk. The DynamicRoutingEngine needs to be as resilient as somehow possible. Therefore I would also suggest in introducing an "updateRoute" function that handles the stopping and restarting of the dynamicRouting engine in all kinds of edge cases (eg processing a new route from the `routingEngine` vs from the `DynamicRoutingEngine` etc). The documentation is by far not clear enough on that matter and ANRs in the middle of driving are as bad as crashes.

I use the SDK 4.13.5.0 in the Android Navigate edition.

**Comment by HERE-SDK-Support-Team** [comment]
**Update** 

Sorry for the inconvenience. The **DynamicRoutingEngine** is currently undergoing a few more bug fixes and improvements. Until further notice, please ignore our comments from above. Instead, for simplicity, the recommended flow to set new routes looks like this:

1. Decide if a new route should be set.
2. If yes, stop the `DynamicRoutingEngine`.
3. Set the new route: `navigator.setRoute(newRoute)`.
4. Start the `DynamicRoutingEngine` with the new route.

Make sure to call these steps outside of the `onBetterRouteFound()` event: Use a local flag and follow the above steps when receiving a new location update - for example, before calling `dynamicRoutingEngine.updateCurrentLocation(..)`.

The documentation and example apps will be updated for the next release(s).

**Comment by HERE-SDK-Support-Team** [comment]
We plan to simplify the API for future releases, but for now, to be on the safe side (in pseudo code), the following code shows how a flag could be used to set the new route during navigation:

```
// Member used as flag.
Route betterRoute = null;

listener = onBetterRouteFound(route) {
   // Decide if route is better.
   ...
   if (betterRoute == null) { // Check the flag.
      betterRoute = route;
   }
}

onRouteProgressUpdated(...) {
   ...
   if (lastMapMatchedLocation != null) {
      if (betterRoute != null) {
         // Follow the step to set a new route. 
         dynamicRoutingEngine.stop();
         navigator.setRoute(betterRoute);
         dynamicRoutingEngine.start(betterRoute, listener);
         betterRoute = null; // reset the flag
      } 
      dynamicRoutingEngine.updateCurrentLocation(lastMapMatchedLocation, routeProgress.sectionIndex);
  }
}
```

The same flag is then expected to be used also when using, for example, `returnToRoute()` or other features that provide a new route. We hope this clarifies the usage of how a flag could be used. In order to be sure it is thread-safe, you can synchronize the flag, but since onRouteProgressUpdated() and onBetterRouteFound() are called on the same thread, the main thread, this should not be necessary.

**Comment by StefanTiess** [comment]
Hi, thank you for looking into it. The Crashes have gotten much better. Unfortunately I still see numerous ANRs in the DynamicRoutingEngine.onStop() method. This is even true for the 4.14.1 SDK Version that has marked the issues in the onStop as resolved in the Release notes.

They all appear in the same moment:
* When a RouteDeviation is detected I call DynamicRoutingEngine.onStop() right before I search for the return route with RoutingEngine.returnToRoute(...)

Is there something I can do to avoid this? Or do I need to wait for an additional fix here?

**Comment by HERE-SDK-Support-Team** [comment]
Hi Stefan, thanks for the feedback. Are you calling `dynamicRoutingEngine.onStop()` in the `onRouteProgressUpdated() `callback?

**Comment by StefanTiess** [comment]
Hi,

i changed it in the meantime to follow completely the "better" route approach and the issues got much better. But still, even with all the proposed fixes in place, ANRs in the dynamicRoutingEngine.onStop remain the most probable cause for a Crash in my app. Are there any additional fixes planed for coming versions to stabilize this feature?

Also, i see many many native crashes exclusively for Adreno-Devices in the libGLESv2_adreno.so. Are you aware of that issue and is there something I can do to mitigate?

Best regards,
Stefan

**Comment by HERE-SDK-Support-Team** [comment]
Thanks again for reporting, @StefanTiess. Are these ANRs happening frequently on specific devices? Have you tried to use the latest HERE SDK version? 

For the crashes in libGLESv2_adreno.so - in which way is this related to the HERE SDK? Do you ave some more information on these crashes? Thanks.

**Comment by HERE-SDK-Support-Team** [comment]
PS: The HERE SDK only supports GLES >= 3.0, see our [minimum requirements](https://developer.here.com/documentation/android-sdk-explore/4.14.3.0/dev_guide/topics/about.html#minimum-requirements).

**Comment by HERE-SDK-Support-Team** [comment]
@StefanTiess Sorry for one more question. We have reopened the ANR case #17310. Since the fix was delivered with 4.14.1 - can you provide us with a crash dump of the new occurrence? Thank you.

**Comment by StefanTiess** [comment]
Hi, sorry for the delay. These are a few samples of the ANR just from the last 24hrs.
[com.cargotour.app_issue_60c6295e590c8daa533d7877c72bd37c_ANR_session_64A1C6A400AA00010E143059EA67EE49_DNE_0_v2_stacktrace.txt](https://github.com/heremaps/here-sdk-examples/files/11934060/com.cargotour.app_issue_60c6295e590c8daa533d7877c72bd37c_ANR_session_64A1C6A400AA00010E143059EA67EE49_DNE_0_v2_stacktrace.txt)
[com.cargotour.app_issue_052718fa4aaf06285ab34cd0eb213ec7_ANR_session_64A23C3B034E00014EBE1C1A6D4566BD_DNE_0_v2_stacktrace.txt](https://github.com/heremaps/here-sdk-examples/files/11934061/com.cargotour.app_issue_052718fa4aaf06285ab34cd0eb213ec7_ANR_session_64A23C3B034E00014EBE1C1A6D4566BD_DNE_0_v2_stacktrace.txt)
[com.cargotour.app_issue_3871147b3f74f16bcbb7b9dbaa85036d_ANR_session_64A23B02035000012A64FC8A2EFD6AA8_DNE_0_v2_stacktrace.txt](https://github.com/heremaps/here-sdk-examples/files/11934063/com.cargotour.app_issue_3871147b3f74f16bcbb7b9dbaa85036d_ANR_session_64A23B02035000012A64FC8A2EFD6AA8_DNE_0_v2_stacktrace.txt)
